Unify weighted mean code

lowener opened the pull request:
Needed for https://github.com/rapidsai/cuml/pull/4428

cjnolet commented on cpp/include/raft/stats/detail/weighted_mean.cuh:
```diff
@@ -18,14 +18,17 @@
 
 #include <raft/cudart_utils.h>
 #include <raft/linalg/coalesced_reduction.hpp>
+#include <raft/linalg/reduce.cuh>
```
These are going to have `hpp` extensions now.

cjnolet commented on cpp/include/raft/stats/weighted_mean.hpp:
```diff
@@ -40,7 +41,35 @@ void rowWeightedMean(
+ * @param stream cuda stream to launch work on
+ */
+template <typename Type>
+void rowSampleWeightedMean(Type* mu,
```
We need a gtest for this as well

lowener commented:
I reworked this PR to unify and clean the weighted mean code:
- I added `weightedMean` which is more general than the col and row functions.
- I fixed `colWeightedMean` documentation which was wrong
- I added a template for IndexType.
- I added a test for the cases not covered by `rowWeightedMean` and `colWeightedMean` (when the input is col-major)

cjnolet commented on cpp/test/stats/weighted_mean.cu:
```diff
@@ -171,29 +172,86 @@ class ColWeightedMeanTest : public ::testing::TestWithParam<WeightedMeanInputs<T
+  void SetUp() override
+  {
+    params = ::testing::TestWithParam<WeightedMeanInputs<T>>::GetParam();
+    raft::random::Rng r(params.seed);
```
Just for future reference, we now have `make_blobs` within RAFT!

cjnolet commented on cpp/test/stats/weighted_mean.cu:
```diff
@@ -171,29 +172,86 @@ class ColWeightedMeanTest : public ::testing::TestWithParam<WeightedMeanInputs<T
+      naiveColWeightedMean(hexp.data(), hin.data(), hweights.data(), rows, cols, false);
+    dexp = hexp;
+
+    // compute ml-prims result
```
We should remove references to ml-prims since it's now legacy.

cjnolet commented on cpp/test/stats/weighted_mean.cu:
```diff
@@ -171,29 +172,86 @@ class ColWeightedMeanTest : public ::testing::TestWithParam<WeightedMeanInputs<T
+                 dweights.data().get(),
+                 cols,
+                 rows,
+                 false,
```
We should add a param for `row_major` as well so we can test that path.

lowener replied:
The `row_major=true` path is already tested through `colWeightedMean` and `rowWeightedMean` so no need to test it twice. I added a comment for the future to make it explicit.

cjnolet replied:
We should test each option in the public API so that the interface isn't making assumptions about the implementation details (e.g. someone changing this in the future might not realize that assumption.)

cjnolet reviewed (changes requested):
I really like the consolidated version. Overall it looks great.

cjnolet reviewed (approved):
LGTM pending CI. Thanks Mickael!

cjnolet commented:
@gpucibot merge
